### PR TITLE
[TT-9138/TT-9907]add syncResourcesWithReload

### PR DESCRIPTION
**Pull request by jeffy-mathew**
## Description

Add retry logic when resource sync fails

## Related Issue
https://tyktech.atlassian.net/browse/TT-9138

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5476/merge
Commit:  
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6001468579)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5476/merge
Commit: de0725feb73a6159de40680ff6d2cd88eb50f751 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6001575902)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5476/merge
Commit: 9c6d0c7b605bf454b294b287ee05e3c732795110 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6009685409)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/5476/merge
Commit: 3c53cfeec15f2b80e8a025826defe9a581115107 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6009689152)

File: gateway/policy.go [hunk 1]
```diff
@@ -17,6 +17,10 @@ import (
 	"github.com/TykTechnologies/tyk/user"
 )
 
+var (
```
Comment (titpetric) [inline gateway/policy.go:20]: lets use `errors.go` for errors, no need to move existing ones yet, but we will eventually :)

File: gateway/server.go [hunk 2]
```diff
@@ -86,6 +86,8 @@ var (
 		// TODO: add ~/.config/tyk/tyk.conf here?
 		"/etc/tyk/tyk.conf",
 	}
+
+	ErrSyncResourceNotKnown = errors.New("unknown resource to sync")
```
Comment (titpetric) [inline gateway/server.go:90]: Same, to `errors.go`

File: gateway/server.go [hunk 3]
```diff
@@ -993,6 +995,30 @@ func (gw *Gateway) DoReload() {
...
+	}
+
+	conf := gw.GetConfig()
+	for i := 1; i <= conf.ResourceReloadAttempts+1; i++ {
+		count, err = syncFunc()
+		if err == nil {
+			return count, nil
+		}
+
+		mainLog.Errorf("Error during syncing %s: %s, attempt count %d", resource, err.Error(), i)
```
Comment (titpetric) [inline gateway/server.go:1015]: maybe skip the log here, log outside

**Review by titpetric — APPROVED** [review]
LGTM, some minor comments

File: gateway/server.go (hunk 3 above)
Comment (jeffy-mathew) [reply gateway/server.go:1015]: I feel like attempt count must be logged. wdyt?

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5476/merge
Commit: 5d908143cef545b867ce532fca24821c1d7c098b 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6023782442)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5476/merge
Commit: fc361d0261c3864321759a5da9d95b93335a00fa 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6023828525)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5476/merge
Commit: 3d857b2d8a07d96bdaf38eb3b9169153ed983fad 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6025497634)

File: config/config.go [hunk 4]
```diff
@@ -1046,6 +1046,17 @@ type Config struct {
...
 	// Skip TLS verification for JWT JWKs url validation
 	JWTSSLInsecureSkipVerify bool `json:"jwt_ssl_insecure_skip_verify"`
+
+	// ResourceSync configures mitigation strategy in case sync fails.
+	ResourceSync ResourceSyncConfig `json:"resource_sync"`
+}
+
+type ResourceSyncConfig struct {
+	// RetryAttempts configures the number of retry attempts before returning on a resource sync.
+	RetryAttempts int `json:"retry_attempts"`
```
Comment (titpetric) [inline config/config.go:1056]: Enabled bool missing

Retry_attempts = 0 (infinity?)

**Review by titpetric — APPROVED** [review]

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5476/merge
Commit: 70cf135385df6a69d20c859ebbdd2ebb034c5388 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/6033113033)

**Pull request merged**